code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -375,4 +388,43 @@ public void cloneConsumingServices(Service fromService, Service toService) {
         consumeMapDao.createServiceLinks(linksToCreate);
     }
 
+    protected String getServiceVIP(Service service, String requestedVip) {
+        if (service.getKind().equalsIgnoreCase(KIND.LOADBALANCERSERVICE.name())
+                || service.getKind().equalsIgnoreCase(KIND.SERVICE.name())) {
+            Subnet vipSubnet = getServiceVipSubnet(service);
+            PooledResourceOptions options = new PooledResourceOptions();
+            if (requestedVip != null) {
+                options.setRequestedItem(requestedVip);
+            }
+            PooledResource resource = poolManager.allocateOneResource(vipSubnet, service, options);
+            if (resource != null) {
+                return resource.getName();
+            }
+        }
+        return null;
+    }
+
+    protected Subnet getServiceVipSubnet(Service service) {
+        Subnet vipSubnet = ntwkDao.addVIPSubnet(service.getAccountId());
```

Comment:
I think this needs a lock around it, or maybe in the DAO code.  So that you don't create two exact subnets in parallel.